src/main/java/org/javarosa/core/model/FormDef.java

```diff
@@ -531,16 +531,25 @@ public void processResultOfAction(TreeReference refSetByAction, String event) {
     }
 
     public boolean isRepeatRelevant(TreeReference repeatRef) {
-        TreeElement repeatNode = mainInstance.resolveReference(repeatRef);
+        boolean relev = true;
 
-        if (repeatNode != null) {
-            return repeatNode.isRelevant();
-        } else {
-            // We are dealing with a repeat that doesn't exist yet so get
-            // relevance from the template
-            TreeElement template = mainInstance.getTemplate(repeatRef);
-            return template.isRelevant();
+        QuickTriggerable qc = dagImpl.getRelevanceForRepeat(repeatRef.genericize());
+        if (qc != null) {
+            relev = (boolean) qc.eval(mainInstance, new EvaluationContext(exprEvalContext, repeatRef));
+        }
+
+        if (relev) {
+            // Must check the template in case of static relevance expressions that wouldn't be in the DAG (e.g. false()).
+            TreeElement templNode = mainInstance.getTemplate(repeatRef);
+            // Check the relevancy of the immediate parent. Unclear whether this is needed because in a normal form-filling
+            // flow the child would not be accessible if its parent is irrelevant.
+            TreeReference parentPath = templNode.getParent().getRef().genericize();
```

Comment:
Did we need this step to pass any of the tests?

Reply:
I couldn’t express a test that failed without it and I tried. But I am feeling a little too nervous to remove it.